internal/clients/nightfall/nightfall_test.go

```diff
@@ -118,33 +117,94 @@ func (n *nightfallTestSuite) TestScan() {
+
+	items := []string{
+		"this is a string",
+		fmt.Sprintf("this has a credit card number %s", exampleCreditCardNumber),
+		"tom cruise!!!!!!",
+	}
+
+	expectedScanReq := createScanReq(detectors, items)
+	expectedTooManyRequestsHTTPResponse := &http.Response{StatusCode: http.StatusTooManyRequests}
+
+	tests := []struct {
+		haveSuccess        bool
+		haveNumReqAttempts int
+		wantResponse       [][]nightfallAPI.ScanResponse
+		desc               string
+	}{
 		{
-			{
-				Fragment: exampleCreditCardNumber,
-				Detector: string(nightfallAPI.CREDIT_CARD_NUMBER),
-				Confidence: nightfallAPI.ScanResponseConfidence{
-					Bucket: string(nightfallAPI.LIKELY),
-				},
-			},
+			haveSuccess:        true,
+			haveNumReqAttempts: 1,
+			wantResponse:       expectedScanResponse,
+			desc:               "success on first scan request",
+		},
+		{
+			haveSuccess:        true,
+			haveNumReqAttempts: 3,
+			wantResponse:       expectedScanResponse,
+			desc:               "success on third scan request attempt",
+		},
+		{
+			haveSuccess:        false,
+			haveNumReqAttempts: 5,
+			wantResponse:       [][]nightfallAPI.ScanResponse(nil),
+			desc:               "failed after max retries",
 		},
-		{},
 	}
+	for _, tt := range tests {
+		mockAPIClient := nightfallapi_mock.NewNightfallAPI(ctrl)
 
-	mockAPIClient.EXPECT().ScanAPI().Return(mockScanAPI)
-	mockScanAPI.EXPECT().ScanPayload(gomock.Any(), gomock.AssignableToTypeOf(expectedScanReq)).Return(expectedScanResp, nil, nil)
+		client := nightfall.Client{
+			APIClient:         mockAPIClient,
+			Detectors:         detectors,
+			InitialRetryDelay: time.Millisecond,
+		}
 
-	resp, err := client.Scan(context.Background(), githublogger.NewDefaultGithubLogger(), items)
-	n.NoError(err, "Received error from Scan")
-	n.Equal(expectedScanResp, resp, "Received incorrect response from Scan")
+		numRetries := tt.haveNumReqAttempts
+		if tt.haveSuccess {
+			numRetries--
+		}
+		for i := 0; i < numRetries; i++ {
+			mockAPIClient.EXPECT().ScanPayload(gomock.Any(), expectedScanReq).
+				Return([][]nightfallAPI.ScanResponse{nil}, expectedTooManyRequestsHTTPResponse, nightfall.ErrMaxScanRetries)
+		}
+		if tt.haveSuccess {
+			mockAPIClient.EXPECT().ScanPayload(gomock.Any(), expectedScanReq).
+				Return(expectedScanResponse, nil, nil)
+		}
+		resp, err := client.Scan(context.Background(), githublogger.NewDefaultGithubLogger(), items)
+		if tt.haveSuccess {
+			n.NoError(err, fmt.Sprintf("Received error from Scan in %s test", tt.desc))
+		} else {
+			n.EqualError(
+				err,
+				nightfall.ErrMaxScanRetries.Error(),
+				fmt.Sprintf("Did not get error from %s test", tt.desc),
+			)
+		}
+		n.Equal(tt.wantResponse, resp, fmt.Sprintf("Received incorrect response from Scan in %s test", tt.desc))
+	}
 }
 
 func createScanReq(dets []*nightfallAPI.Detector, items []string) nightfallAPI.ScanRequest {
 	detectors := make([]nightfallAPI.ScanRequestDetectors, 0, len(dets))
-	for d := range dets {
```

Comment:
this was why we weren't able to read the detectors in the tests

Reply:
ah don't we have an identical function elsewhere?

Reply:
Yea `createScanReq` in `nightfall.go` has the same functionality with a slight difference in implementation. Are you suggesting we make that one public to access in our test?

Reply:
yeah it seems like these 2 things are basically doing the same thing, so we're currently managing this functionality in 2 places - which has led to us each having to solve this bug in the last 24 hours.

If we need that func for our tests, it probably makes sense to just export it

Reply:
gotcha i'll delete the one in the tests